rpmlint/checks/FileDigestCheck.py

```diff
@@ -34,46 +38,90 @@ def _get_digest_group_from_path(self, path):
                 continue
 
             if stat.S_ISLNK(pkgfile.mode) and not self.follow_symlinks_in_group[group]:
-                self.output.add_info('W', pkg, f'{group}-file-symlink', filename)
+                self.output.add_info('E', pkg, f'{group}-file-symlink', filename)
+                result = False
                 continue
 
             pkgfile = pkg.readlink(pkgfile)
             if not pkgfile:
-                self.output.add_info('W', pkg, f'{group}-file-symlink', filename, 'broken symlink')
-                continue
+                self.output.add_info('E', pkg, f'{group}-file-symlink', filename, 'broken symlink')
+                result = False
+        return result
 
-            filename = pkgfile.name
+    def _is_valid_digest(self, pkgfile, digest):
+        algorithm = digest['algorithm']
+        if algorithm == 'skip':
+            return True
 
-            if filename in pkg.ghost_files:
-                self.output.add_info('E', pkg, f'{group}-file-digest-ghost', filename)
-            elif known_digests and filename in known_digests:
-                expected_digests = known_digests[filename]
-                if 'skip' in expected_digests:
-                    pass
-                else:
-                    h = hashlib.new('sha256')
-                    with open(pkgfile.path, 'rb') as fd:
-                        while True:
-                            chunk = fd.read(4096)
-                            if not chunk:
-                                break
-                            h.update(chunk)
-                    hexdigest = h.hexdigest()
-                    if hexdigest not in expected_digests:
-                        self.output.add_info('E', pkg, f'{group}-file-digest-mismatch', filename,
-                                             f'expected:{",".join(expected_digests)}', f'has:{hexdigest}')
-            elif group:
-                self.output.add_info('E', pkg, f'{group}-file-digest-unauthorized', filename)
+        digest_hash = digest['hash']
+        pair = (pkgfile.name, algorithm)
+        if pair not in self.digest_cache:
+            h = hashlib.new(algorithm)
+            with open(pkgfile.path, 'rb') as fd:
+                while True:
+                    chunk = fd.read(4096)
+                    if not chunk:
+                        break
+                    h.update(chunk)
+            self.digest_cache[pair] = h.hexdigest()
+
+        return self.digest_cache[pair] == digest_hash
+
+    def _check_group_digests(self, digest_group, pkg):
+        for filename, digest in digest_group.items():
+            if not self._is_valid_digest(pkg.files[filename], digest):
+                return False
+        return True
+
+    def check_binary(self, pkg):
+        """
+        Check that all files in secured locations are covered by a file digest group
+        in which all files have valid digest.
+        """
+
+        if not self._check_symlinks(pkg):
```

Comment:
There are now two linear passes over all package files, here in `_check_symlinks` and below in `secured_paths = ...`. Could this be an effiency problem in large packages?

Reply:
No, it shouldn't be a problem.